File: loom_engine/loom_parameter_editor/models/polygon_config.py
"""
Data models for polygon configuration.
Supports both regular polygons (defined by parameters) and file-based polygons (spline curves).
"""
from dataclasses import dataclass, field
from typing import List, Optional
from enum import Enum


class PolygonSourceType(Enum):
    """Source type for polygon data."""
    FILE = "file"           # Load from external XML file
    REGULAR = "regular"     # Generate regular polygon from parameters


class PolygonType(Enum):
    """Type of polygon geometry."""
    LINE_POLYGON = "LINE_POLYGON"       # Straight line edges
    SPLINE_POLYGON = "SPLINE_POLYGON"   # Cubic bezier curves


@dataclass
class RegularPolygonParams:
    """Parameters for generating a regular polygon."""
    total_points: int = 4
    internal_radius: float = 0.5
    offset: float = 0.0
    scale_x: float = 1.0
    scale_y: float = 1.0
    rotation_angle: float = 0.0
    trans_x: float = 0.5
    trans_y: float = 0.5
    positive_synch: bool = True
    synch_multiplier: float = 1.0

    def copy(self) -> 'RegularPolygonParams':
        return RegularPolygonParams(
            total_points=self.total_points,
            internal_radius=self.internal_radius,
            offset=self.offset,
            scale_x=self.scale_x,
            scale_y=self.scale_y,
            rotation_angle=self.rotation_angle,
            trans_x=self.trans_x,
            trans_y=self.trans_y,
            positive_synch=self.positive_synch,
            synch_multiplier=self.synch_multiplier
        )


@dataclass
class FileSource:
    """Reference to an external polygon file."""
    folder: str = "polygonSet"
    filename: str = ""
    polygon_type: PolygonType = PolygonType.SPLINE_POLYGON
    filter_type: str = "all"   # "all" | "closed_only"

    def copy(self) -> 'FileSource':
        return FileSource(
            folder=self.folder,
            filename=self.filename,
            polygon_type=self.polygon_type,
            filter_type=self.filter_type
        )


@dataclass
class PolygonSetDef:
    """Definition of a polygon set - either file-based or regular polygon."""
    name: str = "Untitled"
    source_type: PolygonSourceType = PolygonSourceType.FILE
    file_source: Optional[FileSource] = None
    regular_params: Optional[RegularPolygonParams] = None

    def __post_init__(self):
        if self.file_source is None and self.source_type == PolygonSourceType.FILE:
            self.file_source = FileSource()
        if self.regular_params is None and self.source_type == PolygonSourceType.REGULAR:
            self.regular_params = RegularPolygonParams()

    def copy(self) -> 'PolygonSetDef':
        return PolygonSetDef(
            name=self.name,
            source_type=self.source_type,
            file_source=self.file_source.copy() if self.file_source else None,
            regular_params=self.regular_params.copy() if self.regular_params else None
        )


@dataclass
class PolygonSetLibrary:
    """Library containing multiple polygon set definitions."""
    name: str = "MainLibrary"
    polygon_sets: List[PolygonSetDef] = field(default_factory=list)

    def add_polygon_set(self, polygon_set: PolygonSetDef) -> None:
        self.polygon_sets.append(polygon_set)

    def remove_polygon_set(self, name: str) -> bool:
        for i, ps in enumerate(self.polygon_sets):
            if ps.name == name:
                del self.polygon_sets[i]
                return True
        return False

    def get_polygon_set(self, name: str) -> Optional[PolygonSetDef]:
        for ps in self.polygon_sets:
            if ps.name == name:
                return ps
        return None

    def copy(self) -> 'PolygonSetLibrary':
        return PolygonSetLibrary(
            name=self.name,
            polygon_sets=[ps.copy() for ps in self.polygon_sets]
        )

    @classmethod
    def default(cls) -> 'PolygonSetLibrary':
        """Create an empty default library."""
        return cls(name="MainLibrary")

File: loom_swift/Sources/LoomEngine/Config/PolygonConfig.swift
/// How the polygon data is represented in the source file.
public enum PolygonFileType: String, Codable, Sendable {
    case splinePolygon = "SPLINE_POLYGON"
    case linePolygon   = "LINE_POLYGON"
}

/// Parameters for an algorithmically-generated star/regular polygon
/// (`<Source type="regular">` in `polygons.xml`).
public struct RegularPolygonParams: Codable, Sendable {
    /// Number of outer tips; total vertex count = `totalPoints * 2`.
    public var totalPoints:      Int    = 4
    /// Inner radius expressed as a fraction of the outer radius (0.0–1.0).
    public var internalRadius:   Double = 0.5
    /// Rotation offset applied after generation (degrees).
    public var offset:           Double = 0.0
    public var scaleX:           Double = 1.0
    public var scaleY:           Double = 1.0
    public var rotationAngle:    Double = 0.0
    /// Translation; Scala stores centre as (0.5, 0.5) → Swift offset = transX − 0.5.
    public var transX:           Double = 0.5
    public var transY:           Double = 0.5
    public var positiveSynch:    Bool   = true
    public var synchMultiplier:  Double = 1.0
}

/// A named polygon set that references an external XML file **or** is generated
/// algorithmically from `RegularPolygonParams`.
///
/// Corresponds to `<PolygonSet>` in `polygons.xml`.
public struct PolygonSetDef: Codable, Sendable {
    /// The name used by `ShapeDef.polygonSetName` to reference this set.
    public var name: String
    /// Sub-folder inside `polygonSets/` where the file lives.  Usually "polygonSet".
    public var folder: String
    /// Filename of the polygon XML file.  Empty when `regularParams` is non-nil.
    public var filename: String
    /// Controls whether closed or open curves are produced.
    public var polygonType: PolygonFileType
    /// Non-nil when the polygon set is generated from a `<Source type="regular">` element.
    public var regularParams: RegularPolygonParams?

    public init(
        name: String,
        folder: String               = "polygonSet",
        filename: String             = "",
        polygonType: PolygonFileType = .splinePolygon,
        regularParams: RegularPolygonParams? = nil
    ) {
        self.name          = name
        self.folder        = folder
        self.filename      = filename
        self.polygonType   = polygonType
        self.regularParams = regularParams
    }
}

/// The library of polygon set definitions, loaded from `polygons.xml`.
public struct PolygonSetLibrary: Codable, Sendable {
    public var name: String
    public var polygonSets: [PolygonSetDef]

    public init(name: String = "", polygonSets: [PolygonSetDef] = []) {
        self.name = name; self.polygonSets = polygonSets
    }
}

/// Root wrapper matching the `<PolygonConfig>` element.
public struct PolygonConfig: Codable, Sendable {
    public var library: PolygonSetLibrary

    public init(library: PolygonSetLibrary = PolygonSetLibrary()) {
        self.library = library
    }
}
